src/AzureAuth/Commands/Ado/CommandPat.cs

```diff
@@ -60,19 +111,161 @@ private enum OutputMode
+            OutputMode.Base64 => $"{pat.Token.Base64()}\n",
+            OutputMode.Header => $"{pat.Token.AsHeader(AzureAuth.Ado.Authorization.Basic)}\n",
+            OutputMode.HeaderValue => $"{pat.Token.AsHeaderValue(AzureAuth.Ado.Authorization.Basic)}\n",
+            OutputMode.Json => $"{pat.AsJson()}\n",
+            _ => throw new ArgumentOutOfRangeException(nameof(output)),
+        };
+
+        // This option validation could also be handled by using a [Required] attribute,
+        // but for the best user experience we'd like to report multiple issues at once.
+        private bool ValidOptions(ILogger logger)
+        {
+            bool validOptions = true;
+            int scopesCount = this.Scopes?.Length ?? 0;
+
+            if (scopesCount == 0)
+            {
+                logger.LogError($"The {ScopeOption} field is required.");
+                validOptions = false;
+            }
+
+            if (string.IsNullOrEmpty(this.Organization))
+            {
+                logger.LogError($"The {OrganizationOption} field is required.");
+                validOptions = false;
+            }
+
+            if (string.IsNullOrEmpty(this.DisplayName))
+            {
+                logger.LogError($"The {DisplayNameOption} field is required.");
+                validOptions = false;
+            }
+
+            if (string.IsNullOrEmpty(this.PromptHint))
+            {
+                logger.LogError($"The {PromptHintOption} field is required.");
+                validOptions = false;
+            }
+
+            return validOptions;
+        }
+
+        private TokenResult AccessToken(IPublicClientAuth publicClientAuth, CommandExecuteEventData eventData)
+        {
+            return publicClientAuth.Token(
+                AzureAuth.Ado.Constants.AdoParams,
+                this.AuthModes,
+                this.Domain,
+                this.PromptHint,
+                AccessTokenTimeout,
+                eventData);
+        }
+
+        private PatOptions PatOptions()
+        {
+            return new PatOptions
+            {
+                Organization = this.Organization,
+                DisplayName = this.DisplayName,
+                Scopes = this.Scopes,
+            };
+        }
+
+        private IPatClient Client(string accessToken)
+        {
+            var baseUrl = new Uri(string.Join('/', AzureAuth.Ado.Constants.BaseUrl, this.Organization));
+            var credentials = new VssOAuthAccessTokenCredential(accessToken);
+            var connection = new VssConnection(baseUrl, credentials);
+            var tokensHttpClientWrapper = new TokensHttpClientWrapper(connection);
+            return new PatClient(tokensHttpClientWrapper);
+        }
+
+        private IPatCache Cache()
+        {
+            var storageProperties = new StorageCreationPropertiesBuilder(
+                PatStorageParameters.CacheFileName,
+                CacheDirectory)
+            .WithMacKeyChain(

```

Comment:
Do we need to always provide `MacKeyChain` and `WithLinuxKeyring` or we can do a compile time check here and not provide for windows?

Reply:
I think this is a great question! Right now I've just borrowed from existing code, so I'm inclined to keep this as is for now. However, we probably _don't_ need to be providing these values on platforms where they don't make sense. Would love to come back around to this issue in a subsequent PR and resolve it for all places in the code where this is used.

Reply:
I've filed #266 to remind us to do this.